src/CompletionProvider.php

```diff
@@ -169,122 +197,112 @@ public function provideCompletion(PhpDocument $doc, Position $pos): CompletionLi
+            }
+
             foreach ($this->index->getDefinitions() as $fqn => $def) {
-                if (
-                    $def->isGlobal // exclude methods, properties etc.
-                    && (
-                        !$prefix
-                        || (
-                            ((!$namespace || $isFullyQualified) && substr($fqn, 0, $prefixLen) === $prefix)
-                            || (
-                                $namespace
-                                && !$isFullyQualified
-                                && substr($fqn, 0, $namespacedPrefixLen) === $namespacedPrefix
-                            )
-                        )
-                    )
-                    // Only suggest classes for `new`
-                    && (!($node instanceof Node\Expr\New_) || $def->canBeInstantiated)
-                ) {
-                    $item = CompletionItem::fromDefinition($def);
-                    // Find the shortest name to reference the symbol
-                    if ($namespace && ($alias = array_search($def, $aliasedDefs, true)) !== false) {
-                        // $alias is the name under which this definition is aliased in the current namespace
-                        $item->insertText = $alias;
-                    } else if ($namespace && !($prefix && $isFullyQualified)) {
-                        // Insert the global FQN with trailing backslash
-                        $item->insertText = '\\' . $fqn;
-                    } else {
-                        // Insert the FQN without trailing backlash
-                        $item->insertText = $fqn;
+                $fqnStartsWithPrefix = strStartsWith($fqn, $prefix);
+                $fqnContainsPrefix = empty($prefix) || strpos($fqn, $prefix) !== false;
+                if (($def->canBeInstantiated || ($def->isGlobal && !isset($creation))) && $fqnContainsPrefix) {
+                    if ($namespaceDefinition !== null && $namespaceDefinition->name !== null) {
+                        $namespacePrefix = (string)PhpParser\ResolvedName::buildName($namespaceDefinition->name->nameParts, $node->getFileContents());
+
+                        $isAliased = false;
+
+                        $isNotFullyQualified = !($class instanceof Node\QualifiedName) || !$class->isFullyQualifiedName();
+                        if ($isNotFullyQualified) {
+                            foreach ($namespaceImportTable as $alias => $name) {
+                                if (strStartsWith($fqn, $name)) {
+                                    $fqn = $alias;
+                                    $isAliased = true;
+                                    break;
+                                }
+                            }
+                        }
+
+
+                        if (!$isNotFullyQualified && ($fqnStartsWithPrefix || strStartsWith($fqn, $namespacePrefix . "\\" . $prefix))) {
+                            // $fqn = $fqn;
```

Comment:
That could be better, this also reminds me that I was going to inline strStartsWith.